### DISCUSSION TITLE
Sample Rate Issue?

### DISCUSSION THREAD

**Discussion by jonherstein** [dscripka/openWakeWord#103] [Q&A]
I am attempting to run Wyoming Satellite with a USB speakerphone (Jabra 410) and Home Assistant. I can get the Wyoming services running fine on the Pi and the satellite shows up in HA. However, the wake word doesnt seem to be recognized. When I speak, nothing happens. The speaker and mic work, as I have tested with aplay and arecord. One thing, though, is that the mic doesnt record well at 16000Hz (audio is scrambled) so I have adjusted it to 44100Hz. The audio is clear when I play it, but again the wake word doesnt seem to pick up. I noticed openWakeWord has a default sample rate of 16K, so I am wondering if the mismatched sample rates are the problem. If so, how do I change the OWW sample rate in the HA addon? It's not a configuration option. Anything else I should look at? This is SO close to working!!

**Comment by dscripka** [comment]
This is a tricky problem, and you correct that the sample rate mismatch is the problem. The openWakeWord models are only trained on 16khz audio data, so no other sample rate will work correctly. I had considered just doing sample rate conversion within openWakeWord, but there are two main challenges with that approach:

1) It's not very efficient, and there is always some degradation in quality
2) The incoming sample rate must be known

In general, it's much better to let the operating system set the sample rate, and it's odd that your mic doesn't work when set to 16khz. If you are on a Pi, do you know if you are using PulseAudio or ALSA?

As for the HA addon, that would be another potential solution, but I'm not sure if that capability is supported. It might be best to create an issue with HA to see if you can set input device sample rates manually.

**Reply by jonherstein** [reply]
Quick update that confuses me. I assumed the issue was in the recording, but I have listened to the 16000 recorded audio on a different device (just testing arecord out to WAV file) and it sounds fine! It's when I play it back on the Pi it sounds garbled. But I assume this doesnt matter because it's not being played back on the Pi...it's just being fed to OpenWakeWord (I assume). I may try NOT doing local OWW in the meantime, but this is baffling to me!

**Reply by dscripka** [reply]
Ah, that is interesting. If the file sounds fine when recording at 16khz, then it should work fine. It sounds like the issue may be with HA (or the HA addon), somehow the audio isn't getting to openWakeWord correctly.

**Comment by jonherstein** [comment]
Thanks so much for replying, I really appreciate it! I had a feeling this was the issue but was unclear how to fix. It seems like the easiest path is getting the Jabra 410 to sample at 16K, but I’m not not sure how. I think the Pi is using ALSA because I haven’t installed Pulse. Do you think Pulse would help? Anything else you think I should try?Thanks again!On Jan 21, 2024, at 7:18 AM, dscripka ***@***.***> wrote:﻿
This is a tricky problem, and you correct that the sample rate mismatch is the problem. The openWakeWord models are only trained on 16khz audio data, so no other sample rate will work correctly. I had considered just doing sample rate conversion within openWakeWord, but there are two main challenges with that approach:

It's not very efficient, and there is always some degradation in quality
The incoming sample rate must be known

In general, it's much better to let the operating system set the sample rate, and it's odd that your mic doesn't work when set to 16khz. If you are on a Pi, do you know if you are using PulseAudio or ALSA?
As for the HA addon, that would be another potential solution, but I'm not sure if that capability is supported. It might be best to create an issue with HA to see if you can set input device sample rates manually.

**Comment by jonherstein** [comment]
One other idea I had was to take the original audio at 44K, then downsample it to 16K before handing it off to OWW. I don’t know if that would work and be fast enough. And I’m not totally sure how.On Jan 21, 2024, at 9:19 AM, Jon Herstein ***@***.***> wrote:﻿Thanks so much for replying, I really appreciate it! I had a feeling this was the issue but was unclear how to fix. It seems like the easiest path is getting the Jabra 410 to sample at 16K, but I’m not not sure how. I think the Pi is using ALSA because I haven’t installed Pulse. Do you think Pulse would help? Anything else you think I should try?Thanks again!On Jan 21, 2024, at 7:18 AM, dscripka ***@***.***> wrote:﻿
This is a tricky problem, and you correct that the sample rate mismatch is the problem. The openWakeWord models are only trained on 16khz audio data, so no other sample rate will work correctly. I had considered just doing sample rate conversion within openWakeWord, but there are two main challenges with that approach:

It's not very efficient, and there is always some degradation in quality
The incoming sample rate must be known

In general, it's much better to let the operating system set the sample rate, and it's odd that your mic doesn't work when set to 16khz. If you are on a Pi, do you know if you are using PulseAudio or ALSA?
As for the HA addon, that would be another potential solution, but I'm not sure if that capability is supported. It might be best to create an issue with HA to see if you can set input device sample rates manually.
